### ISSUE TITLE
Issue to run RJava, R_HOME not set

### ISSUE THREAD

**Issue by benoitgaudou** [gama-platform/gama#2975]
**Describe the bug**
The environment variable R_HOME has been defined, in the `.bash_profile` file:
```
export R_HOME="/Library/Frameworks/R.framework/Resources/"
```

In GAMA with the rJava extension, when trying to connect GAMA with R software, I have an issue as GAMA / java cannot connect to R, as it cannot see the R_HOME env variable.
To be more precise, `System.getenv("R_HOME")` returns `null`.

I follow instructions in the model, but it does not solve the problem...
```
		write "In the R (RStudio) console, write:";
		write "'install.packages(\"rJava\")' to install the library,";
		write "'library(rJava)', to check the install is correct.";
		write "\nNote on MacOSX, in case of troubles in recent versions of MacOXS, you should first write in a terminal:" color:#red;
		write "R CMD javareconf";
		write "sudo ln -f -s $(/usr/libexec/java_home)/jre/lib/server/libjvm.dylib /usr/local/lib";

```

Any idea ?

**Comment by hqnghi88** [comment]
No idea, long time no touch, so i must find time to redo the flow to check on windows.

**Comment by benoitgaudou** [comment]
Hi Nghi,

Without all the flow, perhaps just could you check that a small Java project in Eclipse and the class 

```
public class TestCollection {
	public static void main(String [] args) {
		
		(System.getenv()).keySet().stream().forEach(x -> System.out.println("" + x + " -- " + System.getenv().get(x)));
		System.out.println(System.getenv("R_HOME"));
	}
}
```

**Comment by hqnghi88** [comment]
Yes on win it shows value set to R_HOME
![image](https://user-images.githubusercontent.com/6105384/80087313-1fddde00-8585-11ea-963d-de9230a5ee7b.png)
But one thing, there is 2 places for win that such kind of variable can be defined, Environment Variable and System Variable, some app understand both, some does not ( did not remember that gama could or not). 
Google say that some use  ~/.bashrc  instead of bash_profile, i have no expe on mac so,...

**Comment by RoiArthurB** [comment]
**TL;DR**

Every bash/profile files won't be loaded except if you're launching it with a command line



**Longer explanation**

For Unix system you **should not** put them in `~/.bashrc` nor `~/.bash_profile` nor even `/etc/profile`. These files are only loaded when you're opening your BASH Terminal (for the 3 of them, and only the last one for any terminal). So, except if you're starting the software by command line, you won't have your variable added.

So the **good place** to do so is to add it in `/etc/environment` and reboot your computer (this variables are loaded on boot, I don't know if it would work on logout/login..). You have to be careful with this file because you can mess up your whole system (if you override an important variable), but it is the place to add an environment variable to your system which will be accessible through any running software !

Also, you can **set temporary variable** (for instance in application shortcut `/gama.desktop`) with the command `env`. This command adds a system level variable but only for the command you're launching after. 

**Examples**

Here're some examples which should add your variable in the code :

```
# Not saved
# Works only with command line or script
env R_HOME=/what/ever/I/want eclipse
``` 

```
# Not saved
# Works only with command line
export R_HOME=/what/ever/I/want 
eclipse
``` 

```
# Saved
# Works with anyway to start eclipse for every users on the computer
echo 'R_HOME=/what/ever/I/want' | sudo tee -a /etc/environment
<reboot your computer>
<start eclipse how you want>
``` 

```
# Saved
# Works with any prompt you'll start only for your user
echo 'export R_HOME=/what/ever/I/want' >> ~/.bashrc
source ~/.bashrc
eclipse
```

**Comment by hqnghi88** [comment]
@sriramab Arthur recall one thing, did you source or restart?
# Saved
# Works with any prompt you'll start only for your user
echo 'export R_HOME=/what/ever/I/want' >> ~/.bashrc
source ~/.bashrc
eclipse

**Comment by sriramab** [comment]
I did for bash_profile, not bashrc. I have to try it.

**Comment by RoiArthurB** [comment]
Refreshing your bash variables won't do anything because they won't be loaded in the software if you don't start it in command line...

**Comment by sriramab** [comment]
The /etc/ method was removed Mac 3-4 years ago. Secondly, this bash environment modification was all working earlier with GAMA  and suddenly stopped working

**Comment by benoitgaudou** [comment]
I tried Arthur's method and does not work on MAcOs :-( this variable is not read anymore as Srirama pointed it out.
A second point is that as a Mac user, I do not think I will run apps from a terminal :-) (but rather on clicking on an icon). 
@sriramab : if you really need to use R and GAMA for your work now, a solution is to run eclipse or GAMA from your terminal (check that the R_HOME variable is set before `echo R_HOME`):
```
>   /Applications/Eclipse-DSL-201909/Eclipse.app/Contents/MacOS/eclipse 
```
or
```
>  /Applications/GAMA1.8_Continuous_withJDK_Mac_03.31.20_1fee0fb.app/Contents/MacOS/Gama 
```

But we will need to find a better solution for random user.
I should dig, perhaps in this post, https://stackoverflow.com/questions/25385934/setting-environment-variables-via-launchd-conf-no-longer-works-in-os-x-yosemite/26311753#26311753

Now I agree with you Srirama: it worked a few years ago. But during the period, MACOSX and eclipse versions has changed ... I read that the environment/system variables management is continuously changing ....

**Comment by hqnghi88** [comment]
@sriramab is this the path that must be set /Library/Frameworks/R.framework/Resources/
can you retry to update the rjava plugin and retry in gama.

**Comment by sriramab** [comment]
@benoitgaudou , please explore this also. I tried .. I get permission errors and I am not daring to make big permission changes with my level of skills :D
https://apple.stackexchange.com/questions/107787/how-to-set-system-wide-environment-variables-on-os-x-mavericks

@hqnghi88 , i uninstalled rJava and install again, then restart GAMA. But ... :( 
and yes ` /Library/Frameworks/R.framework/Resources/` is the path if you have only one version of R, else it is slightly more deeper path.

**Comment by hqnghi88** [comment]
But what? Same error or something?

**Comment by sriramab** [comment]
@hqnghi88 yes same error

I also tried to start console before R_eval like in the code below, but doesnt work either. Here the code
```vbscript
***
* Name: connectR
* Author: srirama
* Description: 
* Tags: Tag1, Tag2, TagN
***/

model connectR

global skills:[RSkill]{
	
	init{
		write command("export R_HOME=/Library/Frameworks/R.framework/Versions/3.6/Resources/ &");
		//write command("/Library/Frameworks/R.framework/Versions/3.6/Resources/Rscript /Users/srirama/Documents/myscript.R");  // this obviously works with an explicit path
		
		do startR;
		write R_eval("runif(5)"); // returns five numbers from uniform distribution.
		
	}
	
	
}


experiment connectR type: gui {
	/** Insert here the definition of the input and output of the model */
	output {
	}
}


```

**Comment by hqnghi88** [comment]
Please confirm again which path do you want to use, in your code later is show a different path

**Comment by sriramab** [comment]
@hqnghi88 as I said, if user has only one R version installed, it will be simple as benoit shows, also what benoit shows will default to latest version installed. If user want to be specific, for example I want to use version 3.6 because of package compatibility issues, then R_HOME can be set to a very specific version of R as I show in the second example. 

So for a general user, the path in Benoit first message is appropriate.

**Comment by hqnghi88** [comment]
No, iwant the exactly what you are using to test a solution.

**Comment by sriramab** [comment]
R_HOME=/Library/Frameworks/R.framework/Versions/3.6/Resources/

**Comment by sriramab** [comment]
![Screen Shot 2020-04-24 at 7 46 55 AM](https://user-images.githubusercontent.com/12668606/80179018-e8604780-85ff-11ea-924b-b071f1c59bb7.png)

**Comment by hqnghi88** [comment]
Ok thanks, can you try update again the plugin?

**Comment by sriramab** [comment]
GAMA crashes :(

**Comment by hqnghi88** [comment]
More detail please, always crash at starting or when launching or when call R

**Comment by sriramab** [comment]
Running the R experiment as in the code above, crashes GAMA.
Other tutorial models work fine

**Comment by hqnghi88** [comment]
I test throughly on windows, and for now (not sure in the past), only the PATH of win will take effect beside the jri libray in gama preference. The system PATH of OS must have an entry point to the executeable file of R (for win it is C:\Program Files\R\R-3.6.3\bin\x64). I dont know if there is the equivalent thing in MacOS , please try to verify, must be where the R.dll located
![image](https://user-images.githubusercontent.com/6105384/80272705-04461500-86f6-11ea-9f5b-73fc9a2cf22f.png)

I dont have to set R_HOME anywhere.

**Comment by sriramab** [comment]
I have downloaded new GAMA, used a new workspace, and reinstalled experimental rJava. GAMA still crashes with R models. Can you perhaps rollback your yesterdays update to the previous version.

**Comment by hqnghi88** [comment]
reverted

**Comment by sriramab** [comment]
Thank you, it does not crash anymore. But still gives error of R_HOME not set. But it is set see screenshot. 

Also please see, the change you made makes gama crash, so may be you can make a note of that.

![Screen Shot 2020-04-25 at 10 56 37 AM](https://user-images.githubusercontent.com/12668606/80275756-a0b2ec00-86e3-11ea-898c-9c75cc7f2052.png)

**Comment by sriramab** [comment]
```
2 occurrences in 2 agents at cycle 0: The R_HOME environment variable is not set. R cannot be run.
in 
in 
in do startR ;
in init  {
	do startR ;
	write self.R_eval(command:'runif(5)') ;
}

in agents Simulation 0, connectR0

```

**Comment by hqnghi88** [comment]
Ok so in the path have that point, can you please retry update the plugin 1 more time?

**Comment by sriramab** [comment]
It crashes again after update

**Comment by hqnghi88** [comment]
OMG, no idea how to fix on macos, it works fluently on win i think i need teamview

**Comment by sriramab** [comment]
be my guest

**Comment by sriramab** [comment]
Ok, so after a whole day of investigation with Nghi, we found that gama.code has no problem, but it is the mac path variable that is not being read by gama.code. 

- Now, if you keep gama.code as is, then issuing the following command on mac terminal and then running R models in GAMA are problem free. But the catch is the command is ineffective after a restart and needs to be reissued. So a solution to make this fix permanent is needed.

```bash
launchctl setenv R_HOME /Library/Frameworks/R.framework/Versions/3.6/Resources
```

- I still wonder why setting R_HOME in bashprofile is not working anymore (suddenly). My only guess is, may be I have many years ago set R_HOME in previous mac methods of etc/path or launchtl , and they stayed on even after further upgrades of the OS. No idea, but seems plausible.

Thank you @hqnghi88  for your time. @benoitgaudou  any ideas to make this launchctl a permanent solution?

**Comment by yosef07** [comment]
Hi guys, I also use a mac and this java plugin is been given hell for several days. But is just as was pointed out by Srirama, tampering with the bash file has change since macs consecutive upgrades. The previous R plugin worked fine, not this rjava after trying several twists and turns in the bash file.

**Comment by AlexisDrogoul** [comment]
@benoitgaudou has probably found the fix for this issue (system wide). As a reminder, fixing it only for the current user should probably involve one of the commands here  : https://support.shotgunsoftware.com/hc/en-us/articles/219042108-Setting-global-environment-variables-on-OS-X

**Comment by benoitgaudou** [comment]
@Srirama Bhamidipati <b.srirama@gmail.com> As it is expressed in the link
provided by Alexis, the way to set permanently the R_HOME system-wide
environment variable is to create in the folder

~/Library/LaunchAgents/ (for the current user, note that this folder is a
hidden folder) or in /Library/LaunchAgents/ (for all users) a file named
 environment.plist and containing the line to launch:



<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "
http://www.apple.com/DTDs/PropertyList-1.0.dtd">

<plist version="1.0">

<dict>

  <key>Label</key>

  <string>my.startup</string>

  <key>ProgramArguments</key>

  <array>

    <string>sh</string>

    <string>-c</string>

    <string>

    launchctl setenv R_HOME /Library/Frameworks/R.framework/Resources/

    </string>

  </array>

  <key>RunAtLoad</key>

  <true/>

</dict>

</plist>


I will update the model 0. Test Connection.gaml in accordance.


Cheers


Benoit

Le sam. 23 mai 2020 à 22:24, Alexis Drogoul <notifications@github.com> a
écrit :

**Comment by yosef07** [comment]
Ty guys, will try this and see if I get the Rjava plugin to run this time.

Cheers

**Comment by yosef07** [comment]
Ty again guys, was able to connect to R!

Cheers,
